internal/core/src/storage/MmapChunkManager.cpp

```diff
@@ -0,0 +1,240 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/MmapChunkManager.h"
+#include "storage/LocalChunkManagerSingleton.h"
+#include <fstream>
+#include <sys/mman.h>
+#include <unistd.h>
+#include "stdio.h"
+#include <fcntl.h>
+#include "log/Log.h"
+
+namespace milvus::storage {
+namespace {
+static constexpr int kMmapDefaultProt = PROT_WRITE | PROT_READ;
+static constexpr int kMmapDefaultFlags = MAP_SHARED;
+};  // namespace
+
+// todo(cqy): After confirming the append parallelism of multiple fields, adjust the lock granularity.
+
+MmapBlock::MmapBlock(const std::string& file_name,
+                     const uint64_t file_size,
+                     BlockType type)
+    : file_name_(file_name), file_size_(file_size), block_type_(type) {
+    allocated_size_.fetch_add(file_size);
+    // create tmp file
+    int fd = open(file_name_.c_str(), O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
+    if (fd == -1) {
+        LOG_ERROR("Failed to open mmap tmp file");
+        return;
+    }
+    // append file size to 'file_size'
+    if (lseek(fd, file_size - 1, SEEK_SET) == -1) {
+        LOG_ERROR("Failed to seek mmap tmp file");
+        return;
+    }
+    if (write(fd, "", 1) == -1) {
+        LOG_ERROR("Failed to write mmap tmp file");
+        return;
+    }
+    // memory mmaping
+    addr_ = static_cast<char*>(
+        mmap(nullptr, file_size, kMmapDefaultProt, kMmapDefaultFlags, fd, 0));
+    if (addr_ == MAP_FAILED) {
+        LOG_ERROR("Failed to mmap");
+        return;
+    }
+    offset_.store(0);
+}
+
+MmapBlock::~MmapBlock() {
+    if (addr_ != nullptr) {
```

Comment:
tricky part is how to handle Fragmentation.
otherwise if we are allocating large disk space, we will always has to grow the disk size because there are no consequent large memories.